include/deal.II/base/task_result.h

```diff
@@ -455,9 +455,18 @@ namespace Threads
   {
     // If we have waited before, then return immediately:
     if (result_is_available)
-      return;
+      {
+        std::lock_guard<std::mutex> lock(mutex);
+
+        // First make clear that the result is no longer available, then
+        // reset the object:
+        result_is_available = false;
+        task_result.reset();
```

Comment:
Resetting the task object here as well as previously done `task.reset()` and switch the order:
```
task.reset();
task_result.reset();
result_is_available = false;
```
So that the memory barrier induced by the `atomic<bool>` variable `result_is_available` makes the changes visible on all threads before `result_is_available` is set to false?

Reply:
This does not sound right. We need to set `result_is_available` to `false` first to make sure other threads don't query `task_result`. We can clear `task` first, but in truth, `task` is already empty because we reset it when we set `result_is_available` to `true`.

So I think this is correct as-is.

I will say that I'm not too concerned about the threading behavior. If you get into a race condition where one thread calls `clear()` and another queries the object's value, you have a bug. We can't help you with that.

Reply:
I see. The new version looks good!